source/common/upstream/cluster_manager_impl.h

```diff
@@ -332,10 +338,115 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
     std::vector<PerPriority> per_priority_update_params_;
   };
 
+  /**
+   * An implementation of an on-demand CDS handle. It forwards the discovery request to the cluster
+   * manager that created the handle.
+   *
+   * It's a protected type, so unit tests can use it.
+   */
+  class OdCdsApiHandleImpl : public OdCdsApiHandle {
+  public:
+    static OdCdsApiHandlePtr create(ClusterManagerImpl& parent, OdCdsApiSharedPtr odcds) {
+      return std::make_unique<OdCdsApiHandleImpl>(parent, std::move(odcds));
+    }
+
+    OdCdsApiHandleImpl(ClusterManagerImpl& parent, OdCdsApiSharedPtr odcds)
+        : parent_(parent), odcds_(std::move(odcds)) {
+      ASSERT(odcds_ != nullptr);
+    }
+
+    ClusterDiscoveryCallbackHandlePtr
+    requestOnDemandClusterDiscovery(const std::string& name,
+                                    ClusterDiscoveryCallbackSharedPtr callback,
+                                    std::chrono::milliseconds timeout) override {
+      return parent_.requestOnDemandClusterDiscovery(odcds_, name, std::move(callback), timeout);
+    }
+
+  private:
+    ClusterManagerImpl& parent_;
+    OdCdsApiSharedPtr odcds_;
+  };
+
   virtual void postThreadLocalClusterUpdate(ClusterManagerCluster& cm_cluster,
                                             ThreadLocalClusterUpdateParams&& params);
 
+  /**
+   * Notifies cluster discovery managers in each worker thread that
+   * the discovery process for the cluster with a passed name has
+   * timed out.
+   */
+  void notifyExpiredDiscovery(const std::string& name);
+
 private:
+  struct ThreadLocalClusterManagerImpl;
+
+  /** A thread-local on-demand cluster discovery manager. It takes care of invoking the discovery
+   *  callbacks in the event of a finished discovery. It does it by installing a cluster lifecycle
+   *  callback that invokes the discovery callbacks when a matching cluster just got added.
+   */
+  class ClusterDiscoveryManager : Logger::Loggable<Logger::Id::upstream> {
```

Comment:
Can you just break this out as its own class? I don't think we gain a lot by burying it here in `private` and it is easier to read this as a standalone entity.

Reply:
Hm, it uses the `ThreadLocalClusterManagerImpl` which is a private type inside `ClusterManagerImpl`. I could break it out, but I would either need to make `ClusterDiscoveryManager` a friend of `ClusterManagerImpl` or refactor the manager in some way to avoid the use of the private type. I'll have a look.